cmd/osbuild-composer/composer.go

```diff
@@ -0,0 +1,268 @@
+	if err != nil {
+		return nil, err
+	}
+
+	artifactsDir, err := c.ensureStateDirectory("artifacts", 0755)
+	if err != nil {
+		return nil, err
+	}
+
+	c.distros, err = distro.NewRegistry(fedora31.New(), fedora32.New(), rhel8.New())
+	if err != nil {
+		return nil, fmt.Errorf("Error loading distros: %v", err)
+	}
+
+	c.rpm = rpmmd.NewRPMMD(path.Join(c.cacheDir, "rpmmd"), "/usr/libexec/osbuild-composer/dnf-json")
+
+	// construct job types of the form osbuild:{arch} for all arches
+	jobTypes := []string{"osbuild"}
+	jobTypesMap := map[string]bool{}
+	for _, name := range c.distros.List() {
+		d := c.distros.GetDistro(name)
+		for _, arch := range d.ListArches() {
+			jt := "osbuild:" + arch
+			if !jobTypesMap[jt] {
+				jobTypesMap[jt] = true
+				jobTypes = append(jobTypes, jt)
+			}
+		}
+	}
+
+	jobs, err := fsjobqueue.New(queueDir, jobTypes)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create jobqueue: %v", err)
+	}
+
+	c.workers = worker.NewServer(c.logger, jobs, artifactsDir)
+
+	return &c, nil
+}
+
+func (c *Composer) InitWeldr(repoPaths []string, weldrListener, localWorkerListener net.Listener) error {
+	archName := common.CurrentArch()
+
+	hostDistro, beta, err := c.distros.FromHost()
+	if err != nil {
+		return err
+	}
+
+	arch, err := hostDistro.GetArch(archName)
+	if err != nil {
+		return fmt.Errorf("Host distro does not support host architecture: %v", err)
+	}
+
+	// TODO: refactor to be more generic
+	name := hostDistro.Name()
+	if beta {
+		name += "-beta"
+	}
+
+	repos, err := rpmmd.LoadRepositories(repoPaths, name)
+	if err != nil {
+		return fmt.Errorf("Error loading repositories for %s: %v", hostDistro.Name(), err)
+	}
+
+	store := store.New(&c.stateDir, arch, c.logger)
+	compatOutputDir := path.Join(c.stateDir, "outputs")
+
+	c.weldr = weldr.New(c.rpm, arch, hostDistro, repos[archName], c.logger, store, c.workers, compatOutputDir)
+
+	c.weldrListener = weldrListener
+	c.localWorkerListener = localWorkerListener
+
+	return nil
+}
+
+func (c *Composer) InitKoji(cert, key string, l net.Listener) error {
+	servers := make(map[string]koji.GSSAPICredentials)
+	for name, creds := range c.config.Koji.Servers {
+		if creds.Kerberos != nil {
+			servers[name] = koji.GSSAPICredentials{
+				Principal: creds.Kerberos.Principal,
+				KeyTab:    creds.Kerberos.KeyTab,
+			}
+		}
+	}
+
+	c.koji = kojiapi.NewServer(c.logger, c.workers, c.rpm, c.distros, servers)
+
+	tlsConfig, err := createTLSConfig(&connectionConfig{
+		CACertFile:     c.config.Koji.CA,
+		ServerKeyFile:  key,
+		ServerCertFile: cert,
+		AllowedDomains: c.config.Koji.AllowedDomains,
+	})
+	if err != nil {
+		return fmt.Errorf("Error creating TLS configuration for Koji API: %v", err)
+	}
+
+	c.kojiListener = tls.NewListener(l, tlsConfig)
+
+	return nil
+}
+
+func (c *Composer) InitRemoteWorkers(cert, key string, l net.Listener) error {
+	tlsConfig, err := createTLSConfig(&connectionConfig{
+		CACertFile:     c.config.Worker.CA,
+		ServerKeyFile:  cert,
```

Comment:
As you probably already know, these lines are in the wrong order. :upside_down_face:

Reply:
Indeed I had seen this yesterday, but forgot to actually push the update. Thanks!